184992788 persist additional referral fields

ttoomey opened the pull request:
### Please squash if merging
[#184992788](https://www.pivotaltracker.com/story/show/184992788)

* store received addresses, phones, and emails on HOH's client record
* If a referral with a matching id exists, update it rather than raising an error
* Persist additional fields on referral and referral_posting as described in spec document
* Update existing client records [per spec](https://github.com/greenriver/hmis-warehouse/pull/3070#issuecomment-1554771068)

questions:
* ~~CustomClientAddress and CustomClientContactPoint require ID fields, I don't know how to populate those~~
* ~~When the references client already exists, should we update the client record from the the referral? Right now that only happens if the client record is new, there's a TODO in the code for this~~
* ~~Currently posting.household_id is not set. Is that right?~~
* ~~Not sure if the address field should be "country" or "county" - seems like a typo somewhere~~

ttoomey committed:
184992788-additional-referral-fields

ttoomey committed:
Merge branch 'release-76' into 184992788-additional-referral-fields

ttoomey committed:
184992788-additional-referral-fields

ttoomey committed:
Merge branch 'release-76' into 184992788-additional-referral-fields

ttoomey committed:
persist referral params

ttoomey committed:
184992788-additional-referral-fields

ttoomey committed:
184992788-additional-referral-fields

ttoomey committed:
cleanups

ttoomey committed:
bugfix

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/ac_hmis/create_referral_job.rb:
```diff
@@ -94,9 +99,65 @@ def create_client(attrs)
+            :use,
+            # :county,  FIXME - spec doc has a "county" field; We don't have that but we do have "country" - maybe there's a typo somewhere?
+          ),
+          # FIXME: unsure what to do with this, use uuid for now
+          AddressID: bogus_id,
```
Thats basically right, generate a uuid. That's what we do for all the HUD "ID" fields. But you can use this func which strips out dashes:
```suggestion
          AddressID: Hmis::Hud::Base.generate_uuid,
```

gigxz commented:
> When the references client already exists, should we update the client record from the the referral? Right now that only happens if the client record is new, there's a TODO in the code for this

Good question, yes, we should update the fields on the existing client. Incoming data is considered "most correct."

Names are particular, though. We should not _overwrite_ the old Name if it differs. Rather, the incoming name should be created as a new Hmis::Hud::CustomClientName marked as `primary`.  The CustomClientName class has an after_save hook that will update the name on the client record. The _old_ primary name should still be retained as a CustomClientName.

Note that the old Client might not _have_ a primary CustomClientName yet. It may just have name fields on the Client record. If that's the case, we should first create a CustomClientName to store the "old" name in. I think adding a function like this (untested) on the Hmis::Hud::Client would make sense 
```
  def build_primary_name
    names.new(
      first: first_name,
      last: last_name,
      middle: middle_name,
      suffix: name_suffix,
      name_data_quality: name_data_quality,
      primary: true,
    )
  end
```
So, once that is saved, you can proceed to save the new one, and mark the old one as non-primary. 

> Currently posting.household_id is not set. Is that right?

Yeah, that's right. That will get set when the posting is accepted into a program.

> Not sure if the address field should be "country" or "county" - seems like a typo somewhere

~~It's county, not country~~ sorry, sorry, don't mind me. The incoming field is "county" because that's the name of their field. We should store it in `CustomClientAddress.district`.

gigxz commented:
@ttoomey questions answered above! let me know if you have any follow-ups. Note I edited my response about the county/country question. THanks!

gigxz commented on drivers/hmis_external_apis/app/jobs/hmis_external_apis/ac_hmis/create_referral_job.rb:
```diff
@@ -94,9 +99,65 @@ def create_client(attrs)
+      }
+      client_address_attrs = params[:addresses].to_a.map do |values|
+        {
+          postal_code: values[:zip],
```
```suggestion
          postal_code: values[:zip],
          district: values[:county],
```

gigxz commented on db/warehouse/migrate/20230517023514_additional_external_referral_fields.rb:
```diff
@@ -0,0 +1,28 @@
+    add_reference postings, :status_note_updated_by, index: {name: "idx_#{postings}_user_2"}
+    add_column postings, :denial_reason, :integer
+    add_column postings, :denial_note, :text
+  end
```
Is referral_result missing?

ttoomey replied:
hah oops, somehow lost that line. Thanks!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/referral_posting.rb:
```diff
@@ -29,6 +32,18 @@ class ReferralPosting < ::HmisExternalApis::HmisExternalApisBase
         assigned_to_other_program_status: 60,
         # closed: 65,
       },
+      referral_result: {
+        client_accepted_referral_result: 1,
+        client_rejected_referral_result: 2,
+        provider_rejected_referral_result: 3,
+      },
```
I don't love having these listed out, since it's a HUD enum that could change in future specifications. Can this be replaced with `::HudLists.referral_result_map.invert`? The `hud_lists.rb` file generated based of `hud_lists.json` that's our source-of-truth for HUD mappings.

ttoomey replied:
Thanks, I fiddled with that map to make the enum labels snake-case like rails expects

ttoomey committed:
Merge branch 'release-76' into 184992788-persist-additional-referral-fields

ttoomey committed:
184992788 address feedback

ttoomey committed:
structure sql

ttoomey committed:
update exiting client records from referrals

Following guidance from https://github.com/greenriver/hmis-warehouse/pull/3070#issuecomment-1554771068

ttoomey committed:
specs, bugfix 184992788

ttoomey commented:
> Good question, yes, we should update the fields on the existing client. Incoming data is considered "most correct."

Thanks for the spec, I've tried to implement client update in this PR including the name reconciliation you describe. For customClientAddresses and contacts, I'm replacing any existing records. Alternatively we could merge address, phones, and emails

ttoomey committed:
bugfix 184992788

ttoomey committed:
linter

gigxz commented:
Looks good to me!

gigxz committed:
Merge branch 'release-76' into 184992788-persist-additional-referral-fields
